docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md

```diff
@@ -0,0 +1,378 @@
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, size_t offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size);
+    virtual int iterator_close(iterator_t it);
+    
+private:
+    Mutex _mutex;
+    size_t *_max_keys;
+    size_t *_num_keys;
+    bool _is_initialized;
+}
+```
+
+### Important data structures
+
+```C++
+// Key metadata
+typedef struct {
+    uint32_t magic;
+    uint16_t metadata_size;
+    uint16_t revision;
+    uint32_t user_flags;
+    uint32_t data_size;
+} key_metadata_t;
+
+// incremental set handle
+typedef struct {
+    char *key;
+    uint32_t create_flags;
+    uint32_t data_size;
+} inc_set_handle_t;
+
+// iterator handle
+typedef struct {
+    void *dir_handle;
+    char *prefix;
+} key_iterator_handle_t;
+```
+
+
+### Initialization and reset
+
+**init function**
+
+Header:  
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
+- Create and take `_mutex`
+- Create the FileSystemStore directory if it doesn't exist
+- Iterate on all files in this directory using `opendir` and `readdir` APIs, and store number of files in `_num_keys` 
+- Set `_is_initialized` to true
+- Release `_mutex`     
+  
+**deinit function**
+
+Header:  
+`virtual int deinit();`
+
+Pseudo code:  
+- if not `_is_initialized` return OK
+- Take `_mutex` 
+- Set `_is_initialized` to false
+- Release `_mutex`   
+
+**reset function**
+
+Header:  
+`virtual int reset();`
+
+Pseudo code:  
+- Take `_mutex`
+- Delete all files under the FileSystemStore directory 
+- Set `_num_keys` to 0
+- Release `_mutex`   
+
+### Core APIs
+
+**set function**
+
+Header:  
+`virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);`
+
+Pseudo code:  
+- if not `_is_initialized` return "not initialized" error
+- Call `set_start` with all fields and a local `set_handle_t` variable
+- Call `set_add_data` with `buffer` and `size`
+- Call `set_finalize`
+- Return OK
+
+**get function**
+
+Header:  
+`virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, size_t offset = 0);`
+
+Pseudo code:  
+- if not `_is_initialized` return "not initialized" error
+- Take `_mutex` 
+- Using the `stat` API, extract file size
+- Call `fopen` API on `key`, read mode to achieve a file handle
+- If failed, release `_mutex` and return "not found" error
+- Using `fread` API, read into a `key_metadata_t` structure
```

Comment:
We shouldn't be calling retarget functions from FileSystemStore. The retarget layer is built entirely on top of the KVStore+FileSystem layer. The retarget layer pulls in a lot of toolchain logic and should generally be avoided until the last minute.

It's a feature for users but not for us, we also give the users the option to bypass the retarget layer to avoid the code cost. But that only works if we don't call the retarget layer internally.

Reply:
I agree. Actually the original design used the file system class APIs (and the file system was passed in the constructor). @dannybenor @offirko What do you think?

Reply:
OK, discussed it internally and we agree. Will revert to the original design.